packages/react/src/components/Text/Text.js

```diff
@@ -0,0 +1,95 @@
+/**
+ * Copyright IBM Corp. 2016, 2018
+ *
+ * This source code is licensed under the Apache-2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import PropTypes from 'prop-types';
+import React, { useContext } from 'react';
+import { TextDirectionContext } from './TextDirectionContext';
+
+function Text({ as: BaseComponent = 'span', children, dir = 'auto', ...rest }) {
+  const context = useContext(TextDirectionContext);
+  const textProps = {};
+  const value = {
+    ...context,
+  };
+
+  if (!context) {
+    textProps.dir = dir;
+    value.direction = dir;
+  } else {
+    const { direction: parentDirection, getTextDirection } = context;
+
+    if (getTextDirection && getTextDirection.current) {
+      const text = getTextFromChildren(children);
+      const override = getTextDirection.current(text);
+
+      if (parentDirection !== override) {
+        textProps.dir = override;
+        value.direction = override;
+      } else if (parentDirection === 'auto') {
+        textProps.dir = override;
+      }
+    } else if (parentDirection !== dir) {
+      textProps.dir = dir;
+      value.direction = dir;
+    } else if (parentDirection === 'auto') {
+      textProps.dir = dir;
+    }
```

Comment:
This was some tricky conditional logic to work with nested `Text` nodes, curious if you have any suggestions @tay1orjones to simplify 🤔 

Could also just always pass in `dir` no matter what the parent is and call it a day. Would simplify this a ton but would mean extra markup. Wasn't sure which way would be better and was curious what others thought would be best. I've been staring at this too long now 😂

Reply:
Yeah that is a tricky block. I don't see a way to simplify. It looks like the complexity just stems from only setting `value.direction` when the `parentDirection` isn't `auto`.

Reply:
> Could also just always pass in dir no matter what the parent is and call it a day.

Also I'm not opposed to this. It feels more explicit via the extra markup but I think either is fine. Would be ideal to get feedback from consumers (or as we implement it ourselves) if this auto nesting functionality is preferred.